### PR TITLE
Switch from qos_event.hpp to event_handler.hpp

### PR DESCRIPTION
**Pull request by sjahr**
### Description
 
Address error happening here https://github.com/ros-planning/moveit2/pull/2020#pullrequestreview-1383635249

```
  Starting >>> moveit_resources_prbt_ikfast_manipulator_plugin
  --- stderr: moveit_ros_occupancy_map_monitor
  In file included from /home/runner/work/moveit2/moveit2/.work/target_ws/src/moveit2/moveit_ros/occupancy_map_monitor/src/occupancy_map_server.cpp:48:
  /opt/ros/rolling/include/rclcpp/rclcpp/qos_event.hpp:18:2: error: This header is obsolete, please include rclcpp/event_handler.hpp instead [-Werror,-W#warnings]
  #warning This header is obsolete, please include rclcpp/event_handler.hpp instead
   ^
  1 error generated.
  gmake[2]: *** [CMakeFiles/moveit_ros_occupancy_map_server.dir/build.make:76: CMakeFiles/moveit_ros_occupancy_map_server.dir/src/occupancy_map_server.cpp.o] Error 1
  gmake[1]: *** [CMakeFiles/Makefile2:204: CMakeFiles/moveit_ros_occupancy_map_server.dir/all] Error 2
  gmake: *** [Makefile:146: all] Error 2
  ---
  Failed   <<< moveit_ros_occupancy_map_monitor [38.5s, exited with code 2]
  Aborted  <<< moveit_resources_prbt_ikfast_manipulator_plugin [22.5s]
```

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit2/blob/main/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by sjahr** [comment]
Waiting until `ros-rolling-rclcpp` version 20.0.0 is available https://github.com/ros2/rclcpp/tags

**Review by JafarAbdi — COMMENTED** [review]
I think this will break main on Humble, we need to check for the distro and include the header accordingly

File: moveit_ros/hybrid_planning/test/hybrid_planning_demo_node.cpp [hunk 1]
```diff
@@ -55,7 +55,11 @@
 #include <rclcpp/node_options.hpp>
 #include <rclcpp/parameter_value.hpp>
 #include <rclcpp/qos.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/hybrid_planning/test/hybrid_planning_demo_node.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/moveit_servo/src/teleop_demo/joystick_servo_example.cpp [hunk 2]
```diff
@@ -48,7 +48,11 @@
 #include <rclcpp/node.hpp>
 #include <rclcpp/publisher.hpp>
 #include <rclcpp/qos.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/moveit_servo/src/teleop_demo/joystick_servo_example.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/moveit_servo/test/publish_fake_jog_commands.cpp [hunk 3]
```diff
@@ -47,7 +47,11 @@
 #include <rclcpp/node.hpp>
 #include <rclcpp/node_options.hpp>
 #include <rclcpp/publisher.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/moveit_servo/test/publish_fake_jog_commands.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/occupancy_map_monitor/src/occupancy_map_server.cpp [hunk 4]
```diff
@@ -45,7 +45,11 @@
 #include <rclcpp/logging.hpp>
 #include <rclcpp/node.hpp>
 #include <rclcpp/publisher.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/occupancy_map_monitor/src/occupancy_map_server.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/perception/semantic_world/src/semantic_world.cpp [hunk 5]
```diff
@@ -47,7 +47,11 @@
 #include <rclcpp/logging.hpp>
 #include <rclcpp/node.hpp>
 #include <rclcpp/publisher.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/perception/semantic_world/src/semantic_world.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/planning/planning_components_tools/src/publish_scene_from_text.cpp [hunk 6]
```diff
@@ -41,7 +41,11 @@
 #include <rclcpp/logging.hpp>
 #include <rclcpp/node.hpp>
 #include <rclcpp/publisher.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/planning/planning_components_tools/src/publish_scene_from_text.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/planning/planning_components_tools/src/visualize_robot_collision_volume.cpp [hunk 7]
```diff
@@ -39,7 +39,11 @@
 #include <rclcpp/executors/multi_threaded_executor.hpp>
 #include <rclcpp/node.hpp>
 #include <rclcpp/publisher.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/planning/planning_components_tools/src/visualize_robot_collision_volume.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/planning/planning_scene_monitor/demos/demo_scene.cpp [hunk 8]
```diff
@@ -42,7 +42,11 @@
 #include <rclcpp/logging.hpp>
 #include <rclcpp/node.hpp>
 #include <rclcpp/publisher.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/planning/planning_scene_monitor/demos/demo_scene.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/planning/planning_scene_monitor/src/current_state_monitor_middleware_handle.cpp [hunk 9]
```diff
@@ -39,7 +39,11 @@
 #include <rclcpp/experimental/buffers/intra_process_buffer.hpp>
 #include <rclcpp/node.hpp>
 #include <rclcpp/qos.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/planning/planning_scene_monitor/src/current_state_monitor_middleware_handle.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/warehouse/src/broadcast.cpp [hunk 10]
```diff
@@ -47,7 +47,11 @@
 #include <rclcpp/node.hpp>
 #include <rclcpp/node_options.hpp>
 #include <rclcpp/publisher.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/warehouse/src/broadcast.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

File: moveit_ros/warehouse/src/save_to_warehouse.cpp [hunk 11]
```diff
@@ -50,7 +50,11 @@
 #include <rclcpp/logging.hpp>
 #include <rclcpp/node.hpp>
 #include <rclcpp/node_options.hpp>
+#if MOVEIT_ROLLING
 #include <rclcpp/event_handler.hpp>
+#elif MOVEIT_HUMBLE
+#include <rclcpp/qos_event.hpp>
+#endif
```
Comment (AndyZe) [inline moveit_ros/warehouse/src/save_to_warehouse.cpp]:
```suggestion
#if MOVEIT_ROLLING
  #include <rclcpp/event_handler.hpp>
#elif MOVEIT_HUMBLE
  #include <rclcpp/qos_event.hpp>
#endif
```

**Comment by AndyZe** [comment]
> I think this will break main on Humble, we need to check for the distro and include the header accordingly

@JafarAbdi we have a dedicated Humble branch now, though. I think it would be OK to break main on Humble.

File: moveit_ros/hybrid_planning/test/hybrid_planning_demo_node.cpp (hunk 1 above)
Comment (JafarAbdi) [reply moveit_ros/hybrid_planning/test/hybrid_planning_demo_node.cpp]: It's automatically formatted with clang-format, CI will fail if I apply this patch

**Comment by JafarAbdi** [comment]
> > I think this will break main on Humble, we need to check for the distro and include the header accordingly
> 
> @JafarAbdi we have a dedicated Humble branch now, though. I think it would be OK to break main on Humble.

That branch is stable and used for binary releases.

Rolling is very unstable (at least that was my experience) to be used by a general user. IMHO, we should always aim to have support for rolling + latest LTS version on main (for people who are ok with building from source to get the latest features)

File: moveit_ros/moveit_servo/CMakeLists.txt [hunk 12]
```diff
@@ -99,7 +99,8 @@ rclcpp_components_register_nodes(${SERVO_COMPONENT_NODE} "moveit_servo::ServoNod
 
 # Add executable for using a controller
 add_library(${SERVO_CONTROLLER_INPUT} SHARED src/teleop_demo/joystick_servo_example.cpp)
-ament_target_dependencies(${SERVO_CONTROLLER_INPUT} ${THIS_PACKAGE_INCLUDE_DEPENDS})
+ament_target_dependencies(${SERVO_CONTROLLER_INPUT} PUBLIC ${THIS_PACKAGE_INCLUDE_DEPENDS})
```
Comment (AndyZe) [inline moveit_ros/moveit_servo/CMakeLists.txt:102]: Why should this be PUBLIC? Nothing else should ever depend on it.

Make the same change at L72, 83, 87, 96, 113, 118, 123, for consistency.

Comment (AndyZe) [reply moveit_ros/moveit_servo/CMakeLists.txt]: I went ahead and pushed this change. Please review! I'm happy with the PR now.

**Review by AndyZe — APPROVED** [review]

File: moveit_ros/moveit_servo/CMakeLists.txt (hunk 12 above)
Comment (JafarAbdi) [reply moveit_ros/moveit_servo/CMakeLists.txt]: I'm sorry I missed your message, I think I tried having it as `PRIVATE` but it failed for humble, which is why I changed them back to Public (the default when nothing is specified?)

File: moveit_common/cmake/moveit_package.cmake [hunk 13]
```diff
@@ -41,6 +41,20 @@ macro(moveit_package)
...
+  endif()
+  add_library(moveit_compatibility INTERFACE)
+  if("$ENV{ROS_DISTRO}" STREQUAL "humble")
+    message(STATUS "Build for ROS2 Humble")
+    target_compile_definitions(moveit_compatibility INTERFACE MOVEIT_HUMBLE)
+  elseif("$ENV{ROS_DISTRO}" STREQUAL "rolling")
+    message(STATUS "Build for ROS2 Rolling")
+    target_compile_definitions(moveit_compatibility INTERFACE MOVEIT_ROLLING)
+  else()
+    message(FATAL_ERROR "Unsupported ROS Distribution: " "$ENV{ROS_DISTRO}")
```
Comment (ChrisThrasher) [inline moveit_common/cmake/moveit_package.cmake:55]:
```suggestion
    message(FATAL_ERROR "Unsupported ROS Distribution: $ENV{ROS_DISTRO}")
```

File: moveit_ros/hybrid_planning/test/CMakeLists.txt [hunk 14]
```diff
@@ -1,10 +1,10 @@
 add_executable(cancel_action cancel_action.cpp)
-ament_target_dependencies(cancel_action ${THIS_PACKAGE_INCLUDE_DEPENDS})
-target_link_libraries(cancel_action ${LIBRARIES})
```
Comment (ChrisThrasher) [inline moveit_ros/hybrid_planning/test/CMakeLists.txt:3]: Was this previously dereferencing a variable that didn't exist?

**Review by ChrisThrasher — APPROVED** [review]
This solution feels like overkill in comparison to disabling `-W#warning` but it's a valid solution nonetheless. Do we plan on removing this extra infrastructure as soon as it's no longer needed? Do we know how soon Humble will get access to this header it needs?

File: moveit_ros/hybrid_planning/test/CMakeLists.txt (hunk 14 above)
Comment (sjahr) [reply moveit_ros/hybrid_planning/test/CMakeLists.txt:3]: It should not  be necessary to link against the target list in this variable

File: moveit_ros/perception/semantic_world/CMakeLists.txt [hunk 15]
```diff
@@ -12,6 +13,6 @@ ament_target_dependencies(moveit_semantic_world
   Eigen3
   Boost
 )
-target_link_libraries(moveit_semantic_world ${OpenCV_LIBRARIES})
+target_link_libraries(moveit_semantic_world PRIVATE moveit_compatibility)
```
Comment (sjahr) [inline moveit_ros/perception/semantic_world/CMakeLists.txt:16]: @JafarAbdi Why did you remove OpenCV_LIBRARIES?

File: moveit_ros/moveit_servo/src/teleop_demo/joystick_servo_example.cpp [hunk 16]
```diff
@@ -91,8 +95,8 @@ enum Button
 
 // Some axes have offsets (e.g. the default trigger position is 1.0 not 0)
 // This will map the default values for the axes
-std::map<Axis, double> AXIS_DEFAULTS = { { LEFT_TRIGGER, 1.0 }, { RIGHT_TRIGGER, 1.0 } };
```
Comment (sjahr) [inline moveit_ros/moveit_servo/src/teleop_demo/joystick_servo_example.cpp:94]: What is the reason for this change?

File: moveit_ros/moveit_servo/CMakeLists.txt (hunk 12 above)
Comment (sjahr) [reply moveit_ros/moveit_servo/CMakeLists.txt]: @AndyZe @JafarAbdi No privacy for ament_target_dependencies :sweat_smile:  https://github.com/ament/ament_cmake/issues/158#issuecomment-475384147

Comment (sjahr) [reply moveit_ros/moveit_servo/CMakeLists.txt]: & because of this we cannot use private in target_link_libraries either :disappointed:

File: moveit_ros/moveit_servo/src/teleop_demo/joystick_servo_example.cpp (hunk 16 above)
Comment (AndyZe) [reply moveit_ros/moveit_servo/src/teleop_demo/joystick_servo_example.cpp:94]: Naming convention, I assume. This is a local variable, not a constant.

Comment (AndyZe) [reply moveit_ros/moveit_servo/src/teleop_demo/joystick_servo_example.cpp:94]: i'll push to fix this. Also, the next variable `button_defaults` is unused.

File: moveit_ros/occupancy_map_monitor/CMakeLists.txt [hunk 17]
```diff
@@ -46,9 +46,11 @@ ament_target_dependencies(moveit_ros_occupancy_map_monitor
 
 add_executable(moveit_ros_occupancy_map_server src/occupancy_map_server.cpp)
 ament_target_dependencies(moveit_ros_occupancy_map_server
-  ${THIS_PACKAGE_INCLUDE_DEPENDS}
+  PUBLIC ${THIS_PACKAGE_INCLUDE_DEPENDS}
 )
-target_link_libraries(moveit_ros_occupancy_map_server moveit_ros_occupancy_map_monitor)
+target_link_libraries(moveit_ros_occupancy_map_server
+  PUBLIC moveit_ros_occupancy_map_monitor
+  PRIVATE moveit_compatibility)
```
[the remark this answers is not in the document]
Comment (JafarAbdi) [reply moveit_ros/occupancy_map_monitor/CMakeLists.txt:53]: I think both should have been private since it's an executable

File: moveit_ros/moveit_servo/src/teleop_demo/joystick_servo_example.cpp (hunk 16 above)
Comment (JafarAbdi) [reply moveit_ros/moveit_servo/src/teleop_demo/joystick_servo_example.cpp:94]: It was clang-tidy related

File: moveit_ros/occupancy_map_monitor/CMakeLists.txt (hunk 17 above)
[the remark this answers is not in the document]
Comment (AndyZe) [reply moveit_ros/occupancy_map_monitor/CMakeLists.txt:53]: OK, I understand now. `moveit_compatibility` isn't externally exposed

**Review by JafarAbdi — APPROVED** [review]

File: moveit_ros/perception/semantic_world/CMakeLists.txt (hunk 15 above)
Comment (JafarAbdi) [reply moveit_ros/perception/semantic_world/CMakeLists.txt]: I thought it was already in the `ament_target_dependencies` should be added there anyway rather than using XXX_libraries

Comment (AndyZe) [reply moveit_ros/perception/semantic_world/CMakeLists.txt:16]: That makes sense. I'll switch back to how you had it.

File: moveit_common/cmake/moveit_package.cmake [hunk 18]
```diff
@@ -41,6 +41,20 @@ macro(moveit_package)
   set(CMAKE_CXX_STANDARD_REQUIRED ON)
   set(CMAKE_CXX_EXTENSIONS OFF)
 
+  if(NOT DEFINED ENV{ROS_DISTRO})
```
Comment (jspricke) [inline moveit_common/cmake/moveit_package.cmake:44]: FYI: this breaks the moveit2_packages: https://github.com/moveit/moveit2_packages/actions/runs/4749763095/jobs/8437320692

The better way would be to check the rclcpp version where this change was introduced.
